### ISSUE TITLE
Feat: More input types

### ISSUE THREAD

**Issue by kiyonlin** [charmbracelet/bubbles#41]
We already have text input.

> A text input field, akin to an \<input type="text"\> in HTML. Supports unicode, pasting, in-place scrolling when the value exceeds the width of the element and the common, and many customization options.

It would be nice if we have some components akin to `<input type="number">`, correspond to `intinput` and `floatinput` in bubbletea. They only accept valid number input.

And what I want most is `boolinput`, which can only let user input `y | n | t | f | 0 | 1 | true | false` etc.

Currently I think of two ways to accomplish this goal：
1. develop independent components
2. Add these type into `textinput`

Btw, all input should know if itself is canceled which is different from zero value(like empty string), because some programs should be terminated if user cancel a prompt.

**Comment by meowgorithm** [comment]
It makes sense to add these. Because of all the complexities of `textinput` like nice cursor blinking and horizontal scrolling it makes the most sense to either:

1. Build these other input types right into textinput
2. Build them on top of `textinput`, with `textinput` embedded in the model
 
I _think_ building them on top of `textinput` makes the most sense so the individual implementations are very isolated and clear.

> Btw, all input should know if itself is canceled which is different from zero value(like empty string), because some programs should be terminated if user cancel a prompt

The notion of cancelling is probably outside the scope of pure text input, but there could be a `prompt` that handles that. What do you think?

* * *

Question about `boolinput`: does it need to be text-based? What I usually do is prompt the user with something like:

> Continue (y/N)

And then I listen for <kbd>y</kbd> to confirm, and all other keys to cancel. If it's a sensitive operation I'll confirm a true response with:

> Are you sure (y/N)

Just a general thought, though. We can still support a text-based bool input.

**Comment by muesli** [comment]
I'd fancy a `Validate` function that gets called before input gets accepted by a `textinput`. We could provide a few common validators (bool, int, ...), but you could also implement one yourself that only accepts very specific inputs.

**Comment by meowgorithm** [comment]
> I don't understand `there could be a prompt that handles that`, could you please elaborate that?

I'm not sure if canceling or accepting the value of a `textinput` belongs in `textinput` itself ([here’s an example](https://gist.github.com/meowgorithm/111bedc380724b51ff5289d23342d586) of how I'd implement it on top of `textinput`). But maybe it makes sense to just put cancel/confirm inside of `textinput`. Totally open to any thoughts on how you might want to do it.
 
> I think `y/N` is the value of `boolinput`, but cancelling is a state of `boolinput`, means "I don't want to answer it".

That make sense. You could also prompt with yes/no/cancel, but regardless we can add a boolean `textinput` too.

**Comment by kiyonlin** [comment]
> But maybe it makes sense to just put cancel/confirm inside of `textinput`.

Bubbletea `textinput` is a prompt "dialog" in a sense imo. So I prefer to put the state inside of `textinput` like you said.

**Comment by obnoxiousnerd** [comment]
Are there any plans for multiline input?

**Comment by meowgorithm** [comment]
@obnoxiousnerd Yes! No timeline but it's definitely something we want to to get in.

**Comment by quenbyako** [comment]
@meowgorithm can we help somehow?

**Comment by obnoxiousnerd** [comment]
I was thinking of ways to implement multiline input. The ideas were:
1. Using multiple `textinput`s
2. Creating a separate component with `textinput` as the base with newlines as delimiter for cursor position in the y direction with overflow handling

What do you guys think?

**Comment by cakir-enes** [comment]
Any updates on `multiline-textinput`s?
Using multiple `textinput` could work until user tries to paste stuff I think. A proper component would be much better.

**Comment by meowgorithm** [comment]
It'll probably make the most sense to add a multiline option to the existing `textinput` because of all the subtleties already handled in there. More than likely a `viewport` will be the most practical way to handle vertical overflow.

We'd prefer to do this one internally because of the significance that `textinput` bears on the Bubble Tea ecosystem, however we can't offer a timeline just yet.

**Comment by maaslalani** [comment]
Hey @kiyonlin happy to share that these features are now possible :)

Multiline input was just added as a new `textarea` bubble in `master`

<img width="600" src="https://stuff.charm.sh/bubbles-examples/textarea.gif />

> And what I want most is boolinput, which can only let user input y | n | yes | no | t | f | 0 | 1 | true | false etc.

This is now possible with `Validate` on the `textinput` which was included in the latest release: https://github.com/charmbracelet/bubbles/releases/v0.12.0

```
func boolValidator(s string) error {
  if s == "y" || s == "n" || s == "t" || s == "t" || s == "0" || s == "1" {
    return nil
  }
  for _, v := range []string{"yes", "no", "true", "false"} {
    if strings.StartsWith(v, s) {
      return nil
    }
  }
  return errors.New("invalid input")
}

textinput.Validate = boolValidator
```
